File: server/app/services/context/narrative.py
"""
Narrative generation module for creating background stories.

This module handles the generation of background narratives from summary
and RAG segments. The narrative provides a coherent context for the LLM
to understand the conversation history.

The narrative generation process:
1. Combine summary content (if available)
2. Combine relevant RAG segments (if available)
3. Use LLM to generate a coherent narrative from the agent's perspective

Narrative Perspective Design:
- Uses internal focalization (agent's viewpoint) rather than external focalization
- The agent is addressed as "You" to enhance immersion and continuity
- This helps the LLM naturally continue the conversation rather than "retell" it
"""

from typing import Optional, Dict, Any, List
from langchain_core.messages import HumanMessage
from langchain_openai import ChatOpenAI
from app.models.llm_config import LLMConfig
from app.services.llm.llm_logger import TokenUsageLogger
from app.logger import logger


class NarrativeService:
    """
    Service for generating background narratives from context components.
    
    This service transforms structured context (summary + segments) into
    a flowing narrative that helps the LLM understand conversation history.
    """
    
    NARRATIVE_PROMPT = """You are a conversation background narrative assistant. Generate a coherent background narrative based on the following information.

{summary_section}

{segments_section}

Integrate the above information into a coherent background narrative with the following requirements:
1. Use second-person perspective (address the agent as "You"). For example: "You have been discussing X with the user. The user mentioned..."
2. Preserve key information and context
3. The narrative should be coherent and flowing, not a simple list
4. Output only the narrative content, without additional explanations

IMPORTANT: The narrative MUST preserve the original language of the conversation.
- If the conversation is in Chinese, write the narrative in Chinese.
- If the conversation is in English, write the narrative in English.
- If the conversation contains multiple languages, the narrative may also contain multiple languages.
- Do NOT translate between languages. Maintain information fidelity."""

    @staticmethod
    def create_chat_model(llm_config: LLMConfig) -> ChatOpenAI:
        """
        Create a ChatOpenAI instance from LLM configuration.
        
        Args:
            llm_config: LLM configuration containing model ID, API key, and base URL
            
        Returns:
            Configured ChatOpenAI instance with moderate temperature for creative output
        """
        return ChatOpenAI(
            model=llm_config.model_id,
            openai_api_base=llm_config.base_url,
            openai_api_key=llm_config.api_key,
            temperature=0.3  # Higher temperature for more creative narrative
        )

    @staticmethod
    async def generate_background_story(
        llm_config: LLMConfig,
        summary: Optional[Dict[str, Any]],
        relevant_segments: List[Dict[str, Any]]
    ) -> Optional[str]:
        """
        Generate a background story from summary and RAG segments.
        
        This method combines the summary and relevant segments into a prompt
        and uses LLM to generate a coherent narrative from the agent's perspective
        (internal focalization).
        
        Args:
            llm_config: LLM configuration for narrative generation
            summary: Summary dictionary with 'version' and 'content' keys
            relevant_segments: List of RAG segment dictionaries with 'content' key
            
        Returns:
            Generated background story text, or None if no input provided
        """
        # Return early if no context available
        if not summary and not relevant_segments:
            return None

        # Build summary section
        summary_section = ""
        if summary:
            summary_section = f"[Conversation Summary]\n{summary['content']}"

        # Build segments section
        segments_section = ""
        if relevant_segments:
            segments_text = "\n".join([
                f"- {seg['content']}"
                for seg in relevant_segments
            ])
            segments_section = f"[Relevant Historical Segments]\n{segments_text}"

        # Format prompt
        prompt = NarrativeService.NARRATIVE_PROMPT.format(
            summary_section=summary_section,
            segments_section=segments_section
        )

        try:
            chat_model = NarrativeService.create_chat_model(llm_config)
            messages = [
                HumanMessage(content=prompt)
            ]

            response = await chat_model.ainvoke(messages, config={"callbacks": [TokenUsageLogger()]})
            background_story = response.content.strip()

            logger.info(f"Generated background story, length: {len(background_story)}")
            return background_story

        except Exception as e:
            logger.error(f"Failed to generate background story: {str(e)}", exc_info=True)
            return None

File: server/internal/service/chat/chatctx/narrative.go
package chatctx

import (
	stdctx "context"
	"fmt"

	"private-buddy-server/internal/model"
	"private-buddy-server/internal/service/llm"

	applogger "private-buddy-server/internal/logger"
)

// cachedNarrativePrompt generates a narrative from summary content only.
// Used for cached narrative generation after summary creation.
const cachedNarrativePrompt = `You are a conversation background narrative assistant. Generate a coherent background narrative based on the summary.

Summary:
%s

Requirements:
1. Use second-person perspective (address the agent as "You"). For example: "You have been discussing X with the user. The user mentioned..."
2. Preserve ALL key information from the summary
3. Transform the summary into a flowing narrative
4. Do NOT add interpretations, judgments, or assumptions
5. Maintain information fidelity

IMPORTANT: The narrative MUST preserve the original language of the conversation.
- If the conversation is in Chinese, write the narrative in Chinese.
- If the conversation is in English, write the narrative in English.
- If the conversation contains multiple languages, the narrative may also contain multiple languages.
- Do NOT translate between languages. Maintain information fidelity.

Output only the narrative content.`

// narrativePrompt generates a narrative from summary + RAG segments.
// Used for legacy real-time narrative generation.
const narrativePrompt = `You are a conversation background narrative assistant. Generate a coherent background narrative based on the following information.

%s

%s

Integrate the above information into a coherent background narrative with the following requirements:
1. Use second-person perspective (address the agent as "You"). For example: "You have been discussing X with the user. The user mentioned..."
2. Preserve key information and context
3. The narrative should be coherent and flowing, not a simple list
4. Output only the narrative content, without additional explanations

IMPORTANT: The narrative MUST preserve the original language of the conversation.
- If the conversation is in Chinese, write the narrative in Chinese.
- If the conversation is in English, write the narrative in English.
- If the conversation contains multiple languages, the narrative may also contain multiple languages.
- Do NOT translate between languages. Maintain information fidelity.`

// GenerateNarrativeFromSummary generates a cached narrative from summary content only.
//
// This is the cached narrative generation method, called in background
// immediately after summary generation. The narrative is stored alongside
// the summary and retrieved at chat time without LLM call.
// Uses TemperatureControlled for creative but controlled output.
func GenerateNarrativeFromSummary(llmConfig *model.LLMConfig, summaryContent string) string {
	if summaryContent == "" {
		return ""
	}

	prompt := fmt.Sprintf(cachedNarrativePrompt, summaryContent)

	chatModel := llm.NewChatModelWithTemperature(llmConfig.BaseURL, llmConfig.APIKey, llmConfig.ModelID, llm.TemperatureControlled)

	result, err := chatModel.Chat(stdctx.Background(), []llm.Message{
		{Role: "user", Content: prompt},
	})
	if err != nil {
		applogger.L.Error("Failed to generate cached narrative", "error", err)
		return ""
	}

	applogger.L.Info("Generated cached narrative from summary", "length", len(result))
	return result
}
